Data/pubtator_to_xml.py

```diff
@@ -0,0 +1,206 @@
+                # Write out each MESH id as own tag
+                for tag_num, ids in enumerate(tag["ID"].split("|")):
+                    # Some ids dont have the MESH:#### form so added case to that
+                    if ":" not in ids:
+                        annt.infons["MESH {}".format(tag_num)] = tag["ID"]
+                    else:
+                        term_type, term_id = ids.split(":")
+                        annt.infons["{} {}".format(term_type, tag_num)] = term_id
+            else:
+                # Some ids dont have the MESH:#### form so added case to that
+                if ":" in tag["ID"]:
+                    term_type, term_id = tag["ID"].split(":")
+                    annt.infons[term_type] = term_id
+                else:
+                    annt.infons["MESH"] = tag["ID"]
+        else:
+            annt.infons["MESH"] = "Unknown"
+
+    location = BioCLocation()
+    location.offset = str(tag["Start"])
+    location.length = str(len(tag["Term"]))
+    annt.locations.append(location)
+    annt.text = tag["Term"]
+    return annt
+
+
+def article_generator(file_lines):
+    """Article Generator
+
+    Returns an article that is a dictionary with the following keywords:
+    Title- an array of the document id and the title string
+    Abstract-  an array of document id, title offset, and abstract string
+    Title_Annot- A filtered array of tags specific to the title
+    Abstract_Annot- A filtered array of tags specific to the abstract
+
+    Keywords:
+    file_lines - this is an array of file lines passed from bioconcepts2pubtator_offsets function
+    """
+    article = {}
+    # title
+    title_heading = file_lines[0].split('|')
+    title_len = len(title_heading[1])
+    article["Title"] = [title_heading[0], title_heading[1]]
+
+    # abstract
+    abstract_heading = file_lines[1]
+    article["Abstract"] = [abstract_heading[0], title_len, abstract_heading[1]]
+
+    # set up the csv reader
+
+    annts = csv.DictReader(file_lines[2:], fieldnames=['Document', 'Start', 'End', 'Term', 'Type', 'ID'], delimiter=str("\t"))
+    sorted_annts = sorted(annts, key=lambda x: x["Start"])
+    article["Title_Annot"] = filter(lambda x: x["Start"] < title_len, sorted_annts)
+    article["Abstract_Annot"] = filter(lambda x: x["Start"] > title_len, sorted_annts)
+
+    return article
+
+
+def bioconcepts2pubtator_offsets(input_file):
+    """Bioconcepts to pubtator
+
+    Yields an article that is a dictionary described in the article generator
+    function.
+
+    Keywords:
+    input_file - the name of the bioconcepts2putator_offset file (obtained from pubtator's ftp site: ftp://ftp.ncbi.nlm.nih.gov/pub/lu/PubTator/)
+    """
+    file_lines = list()
+    with open(input_file, "rb") as f:
+        for line in f:
+            # Convert "illegal chracters" (i.e. < > &) in the main text
+            # into html entities
+            line = cgi.escape(line.rstrip()).encode("ascii", "xmlcharrefreplace")
```

Comment:
This function yields articles that are python representations of articles & tags. XML encoded should occur when writing to XML. Therefore, this should be moved to wherever article dictionaries are being converted to XML. Also, I'd think the BioC package should do this encoding for you?